src/render_helpers/mod.rs

```diff
@@ -127,7 +132,75 @@ pub fn render_to_dmabuf(
         }
     }
 
-    let _sync_point = frame.finish().context("error finishing frame")?;
+    let sync_point = frame.finish().context("error finishing frame")?;
+
+    Ok(sync_point)
+}
+
+pub fn render_to_shm(
+    renderer: &mut GlesRenderer,
+    buffer: &WlBuffer,
+    region: Rectangle<i32, Physical>,
+    size: Size<i32, Physical>,
+    scale: Scale<f64>,
+    transform: Transform,
+    elements: impl Iterator<Item = impl RenderElement<GlesRenderer>>,
+) -> anyhow::Result<SyncPoint> {
+    let buffer_size = smithay::backend::renderer::buffer_dimensions(buffer).unwrap();
+    let offscreen_buffer: GlesRenderbuffer = renderer
+        .create_buffer(Fourcc::Abgr8888, buffer_size)
+        .context("error creating renderbuffer")?;
+    renderer
+        .bind(offscreen_buffer)
+        .context("error binding renderbuffer")?;
+
+    let output_rect = Rectangle::from_loc_and_size((0, 0), size);
+
+    let mut frame = renderer
+        .render(size, transform)
+        .context("error starting frame")?;
+
+    frame
+        .clear([0., 0., 0., 0.], &[output_rect])
+        .context("error clearing")?;
+
+    for element in elements {
+        let src = element.src();
+        let dst = element.geometry(scale);
+
+        if let Some(mut damage) = output_rect.intersection(dst) {
+            damage.loc -= dst.loc;
+            element
+                .draw(&mut frame, src, dst, &[damage])
+                .context("error drawing element")?;
+        }
+    }
+
+    let sync_point = frame.finish().context("error finishing frame")?;
```

Comment:
You can replace this and the `gl.ReadPixels` call below with
```rust
    let mapping =
        render_and_download(renderer, size, scale, fourcc, elements).context("error rendering")?;
    let copy = renderer
        .map_texture(&mapping)
        .context("error mapping texture")?;
```
and then use `ptr::copy_to_nonoverlapping()` from `copy` (which is a `&[u8]`) to `shm_buffer`.

Reply:
I suppose i could. But i don't see why we'd do this. Doesn't this just make the whole operation less efficient, for no discernible gain? If the fact that most of this highlighted block is repeat several times across the file, i did fix that just now.

Reply:
Two things that come to mind:

1. Going through `Renderer` sets up other things in the GL context properly, that may or may not be important.
2. I wouldn't be so sure it would be less efficient. The methods used by `render_and_download()` do a `glReadPixels` into a pixel buffer object, then map it into the main memory. I've seen enough GPU weirdness that I wouldn't be surprised if that is actually faster (or at least not slower) than `glReadPixels` directly into the main memory. Ideally should be profiled though.

Reply:
From what i can tell reading various documentation, i'd imagine you might be right that this operation in itself is faster. However, there is another difference. The current implementation uses a `GlesRenderbuffer`, which from what i can tell is generally more efficient to actually render to. So, which is actually faster? Rendering to a texture, mapping the texture, and then `ptr::copy_nonoverlapping` or rendering to a Renderbuffer and then `glReadPixels`?  I don't know. Neither is excessively slow for me, so i don't think the difference is that significant, but i've pushed a commit which implements the alternative behaviour under `render_to_shm_alt`. It's a drop-in replacement for the other function, feel free to profile it if you want. If it's genuinely faster, then i agree we should use this instead.